janusgraph-core/src/main/java/org/janusgraph/diskstorage/Mutation.java

```diff
@@ -35,10 +36,27 @@
     public Mutation(List<E> additions, List<K> deletions) {
```

Comment:
nitpick: You can redirect calls of this constructor to the new constructor:
```java
public Mutation(List<E> additions, List<K> deletions) {
    this(() -> new ArrayList<E>(additions), () -> new ArrayList<E>(deletions));
}
```

Saves lines and reduces potential for future mistakes ;)

Reply:
I am not sure I get it here. How where do you get converted deletions here?

Reply:
Ah, got it. I didn't check the code

Reply:
I'm in doubt of it because we used to check `Preconditions.checkNotNull(additions)` before starting working with the instance. If we change to above code, we may get NPE in ArrayList construction logic. If we move `checkNotNull` it means we add duplicate `checkNotNull` call.
From another hand `checkNotNull` throws NPE if the instance is `null`. Which is a little bit strange to me. I'm not sure why we really need this check here at all. If an instance is null -> we will get the same `NPE` because of `additions.isEmpty()` call. From another hand, we won't get `NPE` here because we never pass null here. So, I guess, this check was added here just for readability to help other developers that they should not pass null here.
If so, I guess we can remove `checkNotNull` from this place at all and use either `assert additions!=null;` or just add javadoc to these constructors and mention there that passed parameters should not be null.
What do you think about it?

Reply:
I think Assertions are a clean way of handling this. Also, I don't mind having duplicate assertions as much as having duplicate business logic

Reply:
I will change `checkNotNull` checks here to:
```
assert additions!=null;
assert deletions!=null;
```
As they are not enabled by default and thus will be used for testing only. And they will clearly show developers that they should not pass `null` here.

Reply:
Hm. Not sure now. The code looks strange if we are using this logic:
```
public Mutation(List<E> additions, List<K> deletions) {
    this(() -> {
        assert additions!=null;
        return new ArrayList<E>(additions);
    }, () -> {
        assert deletions!=null;
        return new ArrayList<K>(deletions);
    });
}
```
I think the below code is cleaner and also doesn't create an unnecessary suppliers:
```
public Mutation(List<E> additions, List<K> deletions) {
    assert additions!=null;
    assert deletions!=null;
    this.additions = additions.isEmpty()? null : new ArrayList<>(additions);
    this.deletions = deletions.isEmpty()? null : new ArrayList<>(deletions);
}
```
Overall, I see that the whole `Mutation` and usages of `Mutation` should be refracted but I guess in other PRs.

Reply:
What do you think about the updated implementation?

Reply:
I'd prefer to have the logic only once, either in the old constructor or in the new one. Maybe you could replace the new constructor to only do
```java
this(additionsCopySupplier.get(), deletionsCopySupplier.get());
```
But this solution would not check if the suppliers themselves are `null`. On the other hand, the current solution doesn't do this either...
This is just a nitpick. I'm also ok with the current implementation but it would be nicer to keep it as clean as possible.

Reply:
The problem with this solution:
`this(additionsCopySupplier.get(), deletionsCopySupplier.get());` is that it will cause a copy of supplier's returned `ArrayList` to a new `ArrayList` which makes suppliers redundant and removes this optimization.
I think it would be better to leave current optimization as is and make re-factoring of `Mutation` and usages of `Mutation` in other PRs

Reply:
I see